### ISSUE TITLE
WebmanMod with ps3netsrv

### ISSUE THREAD

**Issue by s3anami** [aldostools/webMAN-MOD#642]
Using WebmanMod with ps3netsrv through RetroNAS
Couple of things I have run into
1: There seems to be some limits on the number of items per category and it varies from PSP/PS3/PS1
  This seems to be just a memory limit which is fine. I am just wondering why the limits seem to be limited differently per type of game 
2. Webman does not seem to like subfolders on anything other than PS1 games. 
 If I start filtering games into subfolders they do not show up in WebMan, but do properly within Multiman

**Comment by aldostools** [comment]
1- The free memory in VSH is very limited (maximum 3MB if the memory container 4-bg is selected) or up to 1MB depending of the selected memory profile. The limits per category can be increased for a specific category changing the memory usage in /setup.ps3
More information in: https://github.com/aldostools/webMAN-MOD/wiki/~-Plugin-Memory-Usage

2- Content in remote sub-folders is very limited. You can group the games using folders of single letter (e.g. A, B, C, etc.) or use the same name for file and folder. 

This limitation does not apply for folders stored locally in HDD, NTFS, exFAT or FAT32.

**Comment by s3anami** [comment]
Does sub-folder with letters  only work on certain system types (ie PSP, PS3 ?)
I tried it with PSP and PS3 and was unable to have it successfully recognized. I haven't updated to the latest ps3netsrv released a couple days ago if that fixes it

**Comment by aldostools** [comment]
The subfolders limitation is in webMAN MOD, not in ps3netsrv.

The single letter folder was a feature added almost when I started the project, I haven't revisited that feature recently.
I don't know if it is broken or if it's still working. It is supposed to work for all categories.

**Comment by s3anami** [comment]
oh ok, its broken for me then

**Comment by aldostools** [comment]
A new build of ps3netsrv 20220326 has been released adding support for scanning subdirectories.
It requires to update to the latest commit of webMAN MOD 1.47.38.

**Comment by sairuk** [comment]
ps3netsrv build 20220326 (running on RetroNAS irrelevant its just rpi/linux) with the folder structure

```
/PS2ISO/cd/
/PS2ISO/dvd
```
or (as suggested by op)

```
/PS2ISO/c
/PS2ISO/d
```

I get no results on rescan from webui, if i then drop an iso into the main `PS2ISO` and rescan I get 3 entries for the one iso, so this new routine appears not to work correctly currently.

```
PS2ISO/dummy.iso
```

![image](https://user-images.githubusercontent.com/526372/160711122-648041e6-0e8e-42a4-853a-adb107d9829c.png)

![image](https://user-images.githubusercontent.com/526372/160713104-32966f8b-0188-4f32-ae40-62e21724bb05.png)

Based on the naming of each of the dummy iso entries, i assumed it is enumerating the folders and creating a file entry for each. So with that in mind i removed the subfolders and now i get a single entry for `[] dummy.iso` which is still not right but at least shows the subfolders influence this output.

![image](https://user-images.githubusercontent.com/526372/160713535-ab759113-878b-4be5-9dd6-4e12dfa524dc.png)


@aldostools let me know if i can provide more info to help out at all.

**Comment by s3anami** [comment]
Hi,


The folder support seems to be somewhat broken still.
For me this quote from the wiki does not seem to be true:
"Subfolders are now permitted using ps3netsrv 20220326 or later. Older versions require single letter subfolders (A, B, C, etc.) or sub-folders with the same name of the ISO."

For me, only one letter folders with no special characters are working.

For one of the people contributing to the retronas project, the routine still does not function properly as he commented here:
https://github.com/aldostools/webMAN-MOD/issues/642#issuecomment-1082409163

Thanks,

**Comment by aldostools** [comment]
@sairuk @s3anami please re-download webMAN MOD from the release page or use the online updater available in webMAN Games > webMAN Setup > Update webMAN MOD.

Then report back your results after the test. ps3netsrv build 20220326 was updated to work with the recent builds.

![image](https://user-images.githubusercontent.com/974897/160731820-1fa89587-dbac-42d0-828d-a22fb113beab.png)

**Comment by s3anami** [comment]
it works for me now

**Comment by s3anami** [comment]
folders work fine for me now regardless of folder name now. only thing it breaks is redkey support with ps3 games

**Comment by aldostools** [comment]
> only thing it breaks is redkey support with ps3 games

I'm curious how it could happen. ps3netsrv checks for the .dkey in the "REDKEY" folder when the PS3ISO is mounted. The change implemented in the latest ps3netsrv was only in the directory listing.

Could you provide more details about this issue? e.g. folder structure, screen captures, .dkey file?

**Comment by sairuk** [comment]
i've still got the same behaviour here but I'll dig into it a bit more later on, first time in this space so maybe i'm doing something wrong.

- ps3netsrv 20220326
- webMAN MOD 1.47_38 (md5sum: ee4f04de9d4f3d98c3b030e6ae94b2db)
- HEN 3.0.3
- HFW 4.88.1

tried manual install and ondevice update, same behaviour. 

- Loading HEN
- visiting wMAN MOD page
- Refresh
- Menu does its thing on ps3
- Web page refreshes with dud entries and no sub dirs entries

**Comment by s3anami** [comment]
> > only thing it breaks is redkey support with ps3 games
> 
> I'm curious how it could happen. ps3netsrv checks for the .dkey in the "REDKEY" folder when the PS3ISO is mounted. The change implemented in the latest ps3netsrv was only in the directory listing.
> 
> Could you provide more details about this issue? e.g. folder structure, screen captures, .dkey file?

Folder structure is games in folders A-Z
REDKEY files are all in one folder of the same name at the same level as PS3ISO folder.
Keys are fullset from redumps site
If I don't have keys I get the normal 801 whatever error.
With the folders added the issue is different, the game just black screens on loading

**Comment by s3anami** [comment]
> i've still got the same behaviour here but I'll dig into it a bit more later on, first time in this space so maybe i'm doing something wrong.
> 
> * ps3netsrv 20220326
> * webMAN MOD 1.47_38 (md5sum: ee4f04de9d4f3d98c3b030e6ae94b2db)
> * HEN 3.0.3
> * HFW 4.88.1
> 
> tried manual install and ondevice update, same behaviour.
> 
> * Loading HEN
> * visiting wMAN MOD page
> * Refresh
> * Menu does its thing on ps3
> * Web page refreshes with dud entries and no sub dirs entries

If it matters, I'm on 4.88.2 EvilNat not using HEN at all

**Comment by aldostools** [comment]
Try using the same folder structure in REDKEY.

ps3netsrv basically replaces PS3ISO with REDKEY and .ISO with .dkey

It also looks for. dkey in the same folder of the ISO

**Comment by sairuk** [comment]
Wiped the PS3 back to OFW and started again, same behaviour..

I've took a look over the code and dropped some `printf` in there to try and understand whats going on, I'm not a c dude tho but it doesn't seem to enumerate the dir entries for the subdirs and I don't understand why at the moment. 

As far as I can put together everything is in place at the right versions (webMAN MOD/ps3netsrv)

Current folder layout
```
.
├── BDISO
├── DVDISO
├── GAMES
│   └── [BLES01715] F1 Race Stars™
├── PS2ISO
│   ├── cd
│   │   └── dummy-cd.iso
│   ├── dummy.iso
│   └── dvd
│       └── dummy-dvd.iso
├── PS3ISO
│   ├── Skate 3.dkey
│   ├── Skate 3.ISO
│   └── Skate 3.png
├── PSPISO
└── PSXISO

```

Modded output 
```
ps3netsrv build 20220326 (mod by aldostools)
Path: /data/retronas/ps3/ps3netsrv

Host IP #1: 192.168.200.1:38008
Host IP #2: 192.168.0.176:38008
Waiting for client...
[0] Connection from 192.168.0.134
PATH: /data/retronas/ps3/ps3netsrv/GAMES//
ENTRY: [BLES01715] F1 Race Stars™
DIR:/data/retronas/ps3/ps3netsrv/GAMES//

PATH: /data/retronas/ps3/ps3netsrv/PS3ISO//
ENTRY: Skate 3.png
FILE:/data/retronas/ps3/ps3netsrv/PS3ISO//

PATH: /data/retronas/ps3/ps3netsrv/PS3ISO//
ENTRY: Skate 3.ISO
FILE:/data/retronas/ps3/ps3netsrv/PS3ISO//

PATH: /data/retronas/ps3/ps3netsrv/PS3ISO//
ENTRY: Skate 3.dkey
FILE:/data/retronas/ps3/ps3netsrv/PS3ISO//

PATH: /data/retronas/ps3/ps3netsrv/PS2ISO//
ENTRY: cd
DIR:/data/retronas/ps3/ps3netsrv/PS2ISO//
PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: cd
DIR:/data/retronas/ps3/ps3netsrv/PS2ISO///
PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: dummy.iso
FILE:/data/retronas/ps3/ps3netsrv/PS2ISO///

PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: dvd
DIR:/data/retronas/ps3/ps3netsrv/PS2ISO///
PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: dummy.iso
FILE:/data/retronas/ps3/ps3netsrv/PS2ISO//

PATH: /data/retronas/ps3/ps3netsrv/PS2ISO//
ENTRY: dvd
DIR:/data/retronas/ps3/ps3netsrv/PS2ISO//
PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: cd
DIR:/data/retronas/ps3/ps3netsrv/PS2ISO///
PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: dummy.iso
FILE:/data/retronas/ps3/ps3netsrv/PS2ISO///

PATH: /data/retronas/ps3/ps3netsrv/PS2ISO///
ENTRY: dvd
DIR:/data/retronas/ps3/ps3netsrv/PS2ISO///
```

**Comment by aldostools** [comment]
@sairuk Thank you for your feedback.
I found a bug in the latest ps3netsrv. It looks corrected in this build [ps3netsrv_20220330.zip](https://github.com/aldostools/webMAN-MOD/releases/download/1.47.38/ps3netsrv_20220330.zip)

Btw I also updated webMAN MOD to fix an issue with remote copy.

**Comment by s3anami** [comment]
> Try using the same folder structure in REDKEY.
> 
> ps3netsrv basically replaces PS3ISO with REDKEY and .ISO with .dkey
> 
> It also looks for. dkey in the same folder of the ISO

that fixed it

**Comment by aldostools** [comment]
Thank you again for your feedback.

**Comment by sairuk** [comment]
@aldostools thank you, can confirmed scanning for games in subfolders is fixed in 20220330.

**Comment by aldostools** [comment]
Thank you very much for your confirmation.
